core/store/src/adapter/trie_store.rs

```diff
@@ -167,3 +194,103 @@ pub fn get_key_from_shard_uid_and_hash(shard_uid: ShardUId, hash: &CryptoHash) -
+
+    use crate::adapter::trie_store::TrieStoreAdapter;
+    use crate::NodeStorage;
+
+    const ONE: std::num::NonZeroU32 = match std::num::NonZeroU32::new(1) {
+        Some(num) => num,
+        None => panic!(),
+    };
+
+    #[test]
+    fn test_trie_store_adapter() {
+        let (_tmp_dir, opener) = NodeStorage::test_opener();
+        let store = TrieStoreAdapter::new(opener.open().unwrap().get_hot_store());
+        let shard_uids: Vec<ShardUId> =
+            (0..3).map(|i| ShardUId { version: 0, shard_id: i }).collect();
+        let dummy_hash = CryptoHash::default();
+
+        assert_matches!(
+            store.get(shard_uids[0], &dummy_hash),
+            Err(StorageError::MissingTrieValue(_, _))
+        );
+        {
+            let mut store_update = store.store_update();
+            store_update.increment_refcount_by(shard_uids[0], &dummy_hash, &[0], ONE);
+            store_update.increment_refcount_by(shard_uids[1], &dummy_hash, &[1], ONE);
+            store_update.increment_refcount_by(shard_uids[2], &dummy_hash, &[2], ONE);
+            store_update.commit().unwrap();
+        }
+        assert_eq!(*store.get(shard_uids[0], &dummy_hash).unwrap(), [0]);
+        {
+            let mut store_update = store.store_update();
+            store_update.delete_all_state();
+            store_update.commit().unwrap();
+        }
+        assert_matches!(
+            store.get(shard_uids[0], &dummy_hash),
+            Err(StorageError::MissingTrieValue(_, _))
+        );
+    }
+
+    #[test]
+    fn test_shard_uid_mapping() {
+        let (_tmp_dir, opener) = NodeStorage::test_opener();
+        let store = TrieStoreAdapter::new(opener.open().unwrap().get_hot_store());
+        let parent_shard = ShardUId { version: 0, shard_id: 0 };
+        let child_shard = ShardUId { version: 0, shard_id: 1 };
+        let dummy_hash = CryptoHash::default();
+        // Write some data to `parent_shard`.
+        {
+            let mut store_update = store.store_update();
+            store_update.increment_refcount_by(parent_shard, &dummy_hash, &[0], ONE);
+            store_update.commit().unwrap();
+        }
+        // The data is not visible by child shard, because the mapping has not been set yet.
+        assert_matches!(
+            store.get(child_shard, &dummy_hash),
+            Err(StorageError::MissingTrieValue(_, _))
+        );
+        // Set the shard_uid mapping from `child_shard` to `parent_shard`.
+        {
+            let mut store_update = store.store_update();
+            store_update.set_shard_uid_mapping(child_shard, parent_shard);
+            store_update.commit().unwrap();
+        }
+        // The data is now visible at both `parent_shard` and `child_shard`.
+        assert_eq!(*store.get(child_shard, &dummy_hash).unwrap(), [0]);
+        assert_eq!(*store.get(parent_shard, &dummy_hash).unwrap(), [0]);
+        // Remove the data using `parent_shard` UId.
```

Comment:
Can you also add a case where you remove it by using `child_shard` UId?

Reply:
I'm wondering if we should allow decreasing refcounts using parent shard at all?

Reply:
Once resharded we should not use parent shard.
This test shows what the current behavior is.
Do you mean we should detect such situation and panic, and reflect the panic in this test?

Reply:
The garbage collection will keep using the parent shard id for a few epochs after resharding. Then it will start using the child shard id. Both should work.